src/Disks/IO/ReadBufferFromRemoteFSGather.cpp

```diff
@@ -73,7 +76,51 @@ SeekableReadBufferPtr ReadBufferFromRemoteFSGather::createImplementationBuffer(c
     const auto & object_path = object.remote_path;
 
     size_t current_read_until_position = read_until_position ? read_until_position : object.bytes_size;
-    auto current_read_buffer_creator = [=, this]() { return read_buffer_creator(object_path, current_read_until_position); };
+    std::function<std::unique_ptr<ReadBufferFromFileBase>()> current_read_buffer_creator;
+    current_read_buffer_creator = [=, this]() { return read_buffer_creator(object_path, current_read_until_position, true); };
+
+    if (settings.encryption_settings)
```

Comment:
Thoughy about this part of code again and I think there is no reason for putting `ReadBufferFromEncryptedFile` after `ReadBufferFromRemoteFSGather`, it could be put in between `AsynchronousBoundedReadBuffer` and `ReadBufferFromRemoteFSGather`.
For `CachedOnDiskReadBufferFromFile` it is done here because a different blob must have a different caching buffer, because it is important to cache them separately. But this is not important for decryption.
Why I suggest to move encrypted buffer out of here - because it will help to easier support not only unencrypted cache via `EncryptedObjectStorage`, but also encrypted, which I suggested to do in a previous comment. It will simplify this support because you will be able to add decryption layer without patching settings, but by implicitly wrapping the buffers. The only obstacle here is that we want `AsynchronousBoundedReadBuffer` to lie before the encrypted buffer. This is solvable by adding async layer in `DiskObjectStorage::readFile` and disabling `read_settings.remote_fs_method` from `threadpool` to `read` when passing read settings to `object_storage->readObjects`.
wdyt?